Add GHC.Num.abs spec

renanroberto opened the pull request:
`abs` now returns a non-negative number

renanroberto committed:
Add GHC.Num.abs spec

nikivazou commented:
Thanks!!! Any idea why your tests fails on ghc-9?

renanroberto commented:
@nikivazou to be honest I have no idea =( I was hoping to fix this PR and then use the lessons learned to fix #1869. But without success until the moment. If you have any direction you can give me I would love to jump back into this PR!

andrewthad commented:
The CircleCI logs say:

```
          > tests/pos/abs.hs:5:1: error:
          >     Liquid Type Mismatch
          >     .
          >     The inferred type
          >       VV : GHC.Types.Int
          >     .
          >     is not a subtype of the required type
          >       VV : {VV : GHC.Types.Int | VV >= 0}
          >     .
          >     Constraint id 1
          >   |
          > 5 | f x = abs x
          >   | ^^^^^^^^^^^
```

That doesn't really make sense though. Also, I think that instead of

     GHC.Num.abs :: (GHC.Num.Num a) => a -> {x:a | x >= 0}

It would be better to have

     GHC.Num.abs :: (GHC.Num.Num a) => x:a -> {y:a | (x >= 0) ==> (y = x) && (x < 0) ==> y = -x}

Since the solver will deduce that `y >=0` even if nothing about x is known.

facundominguez commented:
Hello! Is anyone willing to test this in the newest LH?

renanroberto committed:
Attend PR suggestion

renanroberto commented:
> Also, I think that instead of
> 
> ```
>  GHC.Num.abs :: (GHC.Num.Num a) => a -> {x:a | x >= 0}
> ```
> 
> It would be better to have
> 
> ```
>  GHC.Num.abs :: (GHC.Num.Num a) => x:a -> {y:a | (x >= 0) ==> (y = x) && (x < 0) ==> y = -x}
> ```
> 
> Since the solver will deduce that `y >=0` even if nothing about x is known.

Great idea! Added this improvement

> Hello! Is anyone willing to test this in the newest LH?

Not sure how to do this. I've rebased this branch with develop. Should this be enough? Also, I'm having some problem with CircleCI, but I think it is unrelated to the PR

facundominguez committed:
Trivial change to test CircleCI
